### PR TITLE
Fix ProxyDrawables not respecting whether the originals would be drawn

### LINKED ISSUE #1283: ProxyDrawable doesn't respect the original drawable's parent's IsPresent state
If a drawable's parent becomes `IsPresent = false`, its proxied version will still be drawn to the screen.

Resolves ppy/osu#1591.

### PR DESCRIPTION
**Pull request by smoogipoo**
Fixes #1283, fixes ppy/osu#1591.

Alternative to #1505.

Proxying the draw nodes seems to cover every case I can think of without requiring any parent traversal. Drawbacks to this:
* Requires passing a frame counter down through `GenerateDrawNodeSubtree`/`addFromComposite` to determine if the drawnode was generated in the frame from the original drawable.
* Adds around 24 bytes to `ProxyDrawable`.

So the way this all works now is `ProxyDrawable` now generates and adds a `DrawNode` to the subtree, but the only purpose of this specialised `DrawNode` is to proxy the `DrawNode.Draw()` call to the real `DrawNode` as generated by the original `Drawable`.

From:
screen shot 2018-04-10 at 6 53 32 pm

To:
screen shot 2018-04-10 at 6 52 31 pm

### CODE REVIEW

**Comment by smoogipoo** [comment]
I've now made `ProxyDrawable` a private class nested (in a partial class) in `Drawable`. It's exposed as a `Drawable` to consumers of `CreateProxy` now, with the now public members `IsProxy`/`HasProxy` if they need to specifically check for if it's  aproxy

**Comment by Tom94** [comment]
I personally don't mind large files (your IDE should help deal with them easily) and thus would rather be in favor of not `partial`ing `Drawable` at all. Alternatively, if it were to be `partial`d, then a lot more than just `ProxyDrawable` should be in a partial class.

**Comment by peppy** [comment]
This particular case is where the partial part is only used to store a tightly coupled nested class. We do the same in osu! (`BeatmapManager` -> `BeatmaapManagerWorkingBeatmap`).

I'm fine with this usage of partial, but strongly against the same class being split across files. Hopefully @Tom94 can agree to this.

**Comment by Tom94** [comment]
I'm fine with that reasoning as long as we're consistent. How about the nested `HandleInputCache`?

**Comment by peppy** [comment]
We generally only do it for larger classes, but even HandleInputCache is probably large enough to be moved out IMO. At the moment it's awkwardly wedged in the middle.

**Comment by smoogipoo** [comment]
+1 to moving HandleInputCache to a partial class too

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 1]
```diff
@@ -676,73 +676,71 @@ protected override void ApplyDrawNode(DrawNode node)
             base.ApplyDrawNode(node);
         }
 
-        protected virtual bool CanBeFlattened => !Masking;
+        protected virtual bool CanBeFlattened => !Masking && !HasProxy;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: i'm not sure i follow why this was added here. may be worthwhile adding a local comment explaining it.

**Comment by Tom94** [comment]
I am really uncomfortable with adding the complexity of the frame counter and validation along with `DrawNode` cross-references. I wonder if a nicer solution exists, because I'd like to keep the code Update/Draw pipeline as lean as possible.

On second thought I am also starting to doubt whether it makes sense to hide the `Proxy` when the parent is hidden, but _not_ when it is masked away in the first place. That decision seems inconsistent. Since we also argued that partial masking would be impossible, I'd actually start leaning towards the old behaviour of not tying the Proxy to its `Parent`'s state in general.

Lastly, we should also not ignore the possibility that `ProxyDrawable` may not be the optimal mechanism for the thing we're trying to achieve on the osu! side. Are there any alternatives? IIRC we didn't think very long and hard when we decided to add `ProxyDrawable` as a simple way to make approach circles work. If `ProxyDrawable` starts to get used in more complex ways and requires additional overhead we should re-evaluate if it's truly worth it.

**Comment by smoogipoo** [comment]
1.

Re: Counter. I don't think the counter is too offensive, but we _may_ be able to remove that in-exchange for a flag that gets set + reset on validation + on draw respectively. I'll experiment with this if you'd like.

Re complexity. I can't think of any other way to do this other than doing `DrawNode` cross references/validation, since you can't know whether the original `Drawable` would have been drawn until post-DrawNode-generation. Would be open to suggestion, however I've thought about this for many days and have attempted many things.

2.

Re: Masked away. What do you mean by "not when it is masked away"? If the proxied drawable is outside of the proxy's parents' (which is masking) masking area, the proxied drawable will be "masked away" without `IsMaskedAway` being set to true.

E.g. with the following code:

```
new Container
{
    Children = new Drawable[]
    {
        box,
        new Container { Masking = true, Child = box.CreateProxy() }
    }
}
```

The following is possible (yellow is the masking container, red is box's proxy with the original box's position/size, white is the drawn part):
screen shot 2018-04-20 at 3 25 24 pm

Obviously you can imagine that as the box is moved more outside the container, less and less of it will be drawn, until a point where it is completely masked away. This stays true to the definition of `IsMaskedAway` which states "but it may be false, even if the Drawable was masked away" (i.e. in the above scenario, `IsMaskedAway` will be false).

3.

You're free to suggest an alternative for osu!, however this doesn't exclude the fact that `ProxyDrawable`s are inherently broken in their current state.

**Comment by Tom94** [comment]
1. I don't have a better idea in mind; just wanted to make sure the design decision was thought about long and hard.

2. I think you misunderstand. I'm talking about the case where the _original is masked away by the original parent_. I think it is inconsistent to _keep drawing_ the proxy in this case, but to _not keep drawing_ the proxy when the original parent hides the original in some other way than masking (e.g. by `!IsPresent` or `!IsAlive`). Either all of these cases should hide the proxy or none of them, and I favour "none", because "all" would require some sensible form of masking by the original, which we cannot do.

3. I am not fully familiar with the context in which proxies are used within lazer and frankly I do not have the time to follow both lazer _and_ framework development in detail. This makes it hard for me to come up with any specific ideas, so I leave the final decisions to you. Just giving my 2 cents on what I think may be the best course of action _in case there is another way_. IIRC for approach circles proxies were only used to customise draw order without throwing hitobject grouping under the bus; nothing else.

**Comment by smoogipoo** [comment]
2.

* When it is fully masked away by its parent, it's not going to be drawn by the proxy, since the original won't generate draw nodes. This is consistent with `!IsPresent` / `!IsAlive`.
* When it is partially masked away by its parent (or even not masked away at all), the original would be drawn, so will be drawn by the proxy. The only difference is that the proxy is the one that has the final say as to how things appear on screen as it does the final drawnode masking.

**Comment by smoogipoo** [comment]
Your "none" proposal is infeasible and potentially dangerous, as that means you can remove a drawable from the hierarchy and still have it drawn by the proxy, meanwhile it's not getting any updates and may potentially even have no parent.

**Comment by Tom94** [comment]
Still sounds strange to me. If full masking by the original parent has any effect, why wouldn't partial masking do so, too?

Also, I find it to be a weird notion, that moving the original around without affecting how the proxy is ultimately drawn can result in the proxy suddenly becoming invisible. Seems like annoying mental overhead that is required when using proxies. As far as I can tell, as it stands now, proxied drawables are simply the originals inserted at other arbitrary positions within the scene graph without any strings attached. This sounds like the most straightforward model to me.

Could you refresh my mind on what the use case of the new behaviour would be within lazer?

**Comment by Tom94** [comment]
Re "none" proposal being dangerous: Should be trivial to throw an exception in such a case.

**Comment by smoogipoo** [comment]
We can change the original to never have its IsMaskedAway never updated, I wouldn't mind that on a first thought.

There are two immediate use cases for this in lazer:
1. Approach Circles (as previously discussed), where you should be able to fastforward in the editor at any time and cause the original to become `!IsAlive`, and have it not be drawn.
2. Swells (taiko), where they're proxied into an upper layer so they don't underlap the black underlay for the drum on the left. Upon fastforwarding in something like the editor, again, the hitobject will be in an `!IsAlive` state and should be hidden.

So your exception would be essentially "if the original is not drawn, throw an exception"? No way, that will cause so many issues and make limit `ProxyDrawable` to being used in completely static contexts.

**Pull request merged**